pymc4/variational/approximations.py

```diff
@@ -0,0 +1,224 @@
+"""Implements ADVI approximations."""
+from typing import Optional, Union
+from collections import namedtuple
+
+# import arviz as az
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from pymc4 import flow
+from pymc4.coroutine_model import Model
+from pymc4.inference.utils import initialize_sampling_state
+
+tfd = tfp.distributions
+tfb = tfp.bijectors
+ADVIFit = namedtuple("ADVIFit", "approximation, losses")
+
+
+class Approximation(tf.Module):
+    """Base Approximation class."""
+
+    def __init__(self, model: Optional[Model] = None, random_seed: Optional[int] = None):
+        if not isinstance(model, Model):
+            raise TypeError(
+                "`fit` function only supports `pymc4.Model` objects, but you've passed `{}`".format(
+                    type(model)
+                )
+            )
+
+        self.model = model
+        self._seed = random_seed
+        self.state, self.deterministic_names = initialize_sampling_state(model)
+        if not self.state.all_unobserved_values:
+            raise ValueError(
+                f"Can not calculate a log probability: the model {model.name or ''} has no unobserved values."
+            )
+
+        self.unobserved_keys = self.state.all_unobserved_values.keys()
+        self.target_log_prob = self._build_logfn()
+        self.approx = self._build_posterior()
+
+    def _build_logfn(self):
+        """Build vectorized logp function."""
+
+        @tf.function(autograph=False)
+        def logpfn(*values, **kwargs):
+            if kwargs and values:
+                raise TypeError("Either list state should be passed or a dict one")
+            elif values:
+                kwargs = dict(zip(self.unobserved_keys, values))
+            st = flow.SamplingState.from_values(kwargs)
+            _, st = flow.evaluate_model_transformed(self.model, state=st)
+            return st.collect_log_prob()
+
+        def vectorize_logp_function(logpfn):
+            def vectorized_logpfn(*q_samples):
+                return tf.vectorized_map(lambda samples: logpfn(*samples), q_samples)
+
+            return vectorized_logpfn
+
+        return vectorize_logp_function(logpfn)
+
+    def _build_posterior(self):
+        raise NotImplementedError
+
+    def flatten_view(self):
+        """Flattened view of the variational parameters."""
+        pass
+
+    def sample(self, n):
+        """Generate samples from posterior distribution."""
+        q_samples = dict(zip(self.unobserved_keys, self.approx.sample(n, seed=self._seed)))
```

Comment:
repetetive samples will be the same

Reply:
convert sampled free vars to actual samples in the model like trace fn in MCMC

Reply:
Resolved in a1d9679